packages/foam-core/src/model/workspace.ts

```diff
@@ -0,0 +1,401 @@
+   */
+  private backlinks: { [key: string]: Connection[] } = {};
+  /**
+   * List of disposables to destroy with the workspace
+   */
+  disposables: IDisposable[] = [];
+
+  exists(uri: URI) {
+    return FoamWorkspace.exists(this, uri);
+  }
+  list() {
+    return FoamWorkspace.list(this);
+  }
+  get(uri: URI) {
+    return FoamWorkspace.get(this, uri);
+  }
+  find(uri: URI) {
+    return FoamWorkspace.find(this, uri);
+  }
+  set(resource: Resource) {
+    return FoamWorkspace.set(this, resource);
+  }
+  delete(uri: URI) {
+    return FoamWorkspace.delete(this, uri);
+  }
+
+  resolveLink(note: Note, link: NoteLink) {
+    return FoamWorkspace.resolveLink(this, note, link);
+  }
+  resolveLinks(keepMonitoring: boolean = false) {
+    return FoamWorkspace.resolveLinks(this, keepMonitoring);
+  }
+  getAllConnections() {
+    return FoamWorkspace.getAllConnections(this);
+  }
+  getConnections(uri: URI) {
+    return FoamWorkspace.getConnections(this, uri);
+  }
+  getLinks(uri: URI) {
+    return FoamWorkspace.getLinks(this, uri);
+  }
+  getBacklinks(uri: URI) {
+    return FoamWorkspace.getBacklinks(this, uri);
+  }
+
+  dispose(): void {
+    this.onDidAddEmitter.dispose();
+    this.onDidDeleteEmitter.dispose();
+    this.onDidUpdateEmitter.dispose();
+    this.disposables.forEach(d => d.dispose());
+  }
+
+  public static resolveLink(
+    workspace: FoamWorkspace,
+    note: Note,
+    link: NoteLink
+  ): URI {
+    let targetUri: URI | null = null;
+    switch (link.type) {
+      case 'wikilink':
+        const definitionUri = note.definitions.find(
+          def => def.label === link.slug
+        )?.url;
+        if (isSome(definitionUri)) {
+          targetUri = parseUri(note.uri, definitionUri);
+        } else {
+          targetUri =
+            FoamWorkspace.find(workspace, link.slug, note.uri)?.uri ??
+            placeholderUri(link.slug);
+        }
+        break;
+
+      case 'link':
+        targetUri = parseUri(note.uri, link.target);
+        break;
+    }
+    if (isPlaceholder(targetUri)) {
+      // we can only add placeholders when links are being resolved
+      workspace = FoamWorkspace.set(workspace, {
+        type: 'placeholder',
+        uri: targetUri,
+      });
+    }
+    return targetUri;
+  }
+
+  /**
+   * Computes all the links in the workspace, connecting notes and
+   * creating placeholders.
+   *
+   * @param workspace the target workspace
+   * @param keepMonitoring whether to recompute the links when the workspace changes
+   * @returns the resolved workspace
+   */
+  public static resolveLinks(
+    workspace: FoamWorkspace,
+    keepMonitoring: boolean = false
+  ): FoamWorkspace {
+    workspace.links = {};
+    workspace.backlinks = {};
+    workspace.placeholders = {};
+
+    workspace = Object.values(workspace.list()).reduce(
+      (w, resource) => FoamWorkspace.addLinksForResource(w, resource),
+      workspace
+    );
+    if (keepMonitoring) {
```

Comment:
This is the part where you said:

> it is possible to keep the connection graph up to date,
> Alternatively the graph can be recomputed when needed

right?

Reply:
yup, that's right

Reply:
do you think this is clear enough in terms of API? I was thinking of otherwise having an options object, something like
`resolveLinks({ keepMonitoring: true })`
at the moment this would be the only param, but who knows if in the future there might be more. more importantly now though, it's that the API feels more explicit then `resolveLinks(true)`

Reply:
Yeah I agree that something like
```
resolveLinks(workspace, { keepMonitoring: true })
```
is clearer and more easily extensible.